02-peer-protocol.md

```diff
@@ -560,6 +1038,302 @@ ones it has sent so it can use them should they be requested by
+   1. type: 1 (`channel_type`)
+   2. data:
+        * [`...*byte`:`type`]
+
+Rationale and Requirements are the same as for [`open_channel`](#the-open_channel-message),
+with the following additions.
+
+#### Requirements
+
+If nodes have negotiated `option_dual_fund`:
+  - the opening node:
+    - MUST NOT send `open_channel`
+
+The sending node:
+  - MUST set `funding_feerate_perkw` to the feerate for this transaction
+
+The receiving node:
+  - MAY fail the negotiation if:
+    - the `locktime` is unacceptable
+    - the `funding_feerate_perkw` is unacceptable
+
+#### Rationale
+
+`zerod_channel_id` for the `open_channel2` MUST be derived using a zero-d out
+basepoint for the peer's revocation basepoint.  This allows the peer to
+return channel-assignable errors before the *accepter*'s revocation
+basepoint is known.
+
+`funding_feerate_perkw` indicates the fee rate that the opening node will
+pay for the funding transaction in satoshi per 1000-weight, as described
+in [BOLT-3, Appendix F](03-transactions.md#appendix-f-dual-funded-transaction-test-vectors).
+
+`locktime` is the locktime for the funding transaction.
+
+The receiving node, if the `locktime` or `feerate_funding_perkw` is considered
+out of an acceptable range, may fail the negotiation. However, it is
+recommended that the *accepter* permits the channel open to proceed
+without their participation in the channel's funding.
+
+Note that `open_channel`'s `channel_reserve_satoshi` has been omitted.
+Instead, the channel reserve is fixed at 1% of the total channel balance
+(`open_channel2`.`funding_satoshis` + `accept_channel2`.`funding_satoshis`)
+rounded down to the nearest whole satoshi or the `dust_limit_satoshis`,
+whichever is greater.
+
+Note that `push_msat` has been omitted.
+
+### The `accept_channel2` Message
+
+This message contains information about a node and indicates its
+acceptance of the new channel.
+
+1. type: 65 (`accept_channel2`)
+2. data:
+    * [`channel_id`:`zerod_channel_id`]
+    * [`u64`:`funding_satoshis`]
+    * [`u64`:`dust_limit_satoshis`]
+    * [`u64`:`max_htlc_value_in_flight_msat`]
+    * [`u64`:`htlc_minimum_msat`]
+    * [`u32`:`minimum_depth`]
+    * [`u16`:`to_self_delay`]
+    * [`u16`:`max_accepted_htlcs`]
+    * [`point`:`funding_pubkey`]
+    * [`point`:`revocation_basepoint`]
+    * [`point`:`payment_basepoint`]
+    * [`point`:`delayed_payment_basepoint`]
+    * [`point`:`htlc_basepoint`]
+    * [`point`:`first_per_commitment_point`]
+    * [`accept_tlvs`:`tlvs`]
+
+1. `tlv_stream`: `accept_tlvs`
+2. types:
+   1. type: 0 (`upfront_shutdown_script`)
+   2. data:
+       * [`...*byte`:`shutdown_scriptpubkey`]
+   1. type: 1 (`channel_type`)
+   2. data:
+        * [`...*byte`:`type`]
+
+Rationale and Requirements are the same as listed above,
+for [`accept_channel`](#the-accept_channel-message) with the following
+additions.
+
+#### Requirements
+
+The accepting node:
+    - MAY respond with a `funding_satoshis` value of zero.
```

Comment:
We just finalize the v2 protocol, with an empty `tx_signatures` from the accepter.

> this would conflict with the requirement that tx_signatures has a nonempty witness_stack

Can you point me to which requirement that is?

> Also, could the accepter still add inputs/outputs to the funding tx without contributing to the funding output?

Yes, that should be accepted!

Reply:
> > this would conflict with the requirement that tx_signatures has a nonempty witness_stack
> 
> Can you point me to which requirement that is? Note the distinction between `witnesses` and `witness_stack`: you are allowed to send 0 witnesses, but if you send a witness, then its `witness_stack` must be non-empty.

Ah, the namings are ambiguous (I mentioned this is a previous review).  From "The `tx_signatures` Message":

```
1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witness_stack`]

1. subtype: `witness_stack`
2. data:
    * [`u16`:`num_input_witness`]
    * [`num_input_witness*witness_element`:`witness_element`]
```

Notice how both the subtype *and* the element in `tx_signatures` are called `witness_stack`?  This makes it unclear what the requirement below refers to:

```
The receiving node:
  - MUST fail the negotiation if:
  ...
    - the message contains an empty `witness_stack`
```

Should we fail if the element `witness_stack` is empty (i.e. `num_witnesses` is 0), or if the subtype `witness_stack` is empty (i.e. `num_input_witness` is 0).